syntax = "proto3";

package officepb;
option go_package = "pb";

message User {
  string name = 1;
  string token = 2;
  string bio = 3;
}

message Document {
  string name = 1;
  Ctx context = 2;
  string content = 3;
}

message Ctx {
  map<string, string> vars = 1;
  repeated Expr exprs = 2;
}

message Expr {
  string name = 1;
  string expr = 2;
}

message RegisterRequest {
  string name = 1;
}

message RegisterResponse {
  User user = 1;
}

message ListRequest {
  int32 offset = 1;
  int32 limit = 2;
}

message ListResponse {
  repeated string usernames = 1;
}

service Users {
  rpc Register(RegisterRequest) returns (RegisterResponse) {}
  rpc List(ListRequest) returns (ListResponse) {}
}